CMakeLists.txt

```diff
@@ -131,39 +139,128 @@ if (LLVM_LINK_LLVM_DYLIB)
     set(LLVM_LIBS LLVM)
 else()
     llvm_map_components_to_libnames(LLVM_LIBS
-      ${LLVM_TARGETS_TO_BUILD} support target option
+      ${LLVM_TARGETS_TO_BUILD} target option
     )
 endif()
 
-# TODO(cmake): With newer cmake there should be a better way to do this.
-function(mk_clang_libs out)
-    if (CLANG_LINK_CLANG_DYLIB)
-        set(${out} clang-cpp PARENT_SCOPE)
-    else()
-        set(${out} ${ARGN} PARENT_SCOPE)
-    endif()
-endfunction()
+if (MLIR_LINK_MLIR_DYLIB)
+    set(MLIR_LIBS MLIR)
+else()
+    get_property(MLIR_DIALECT_LIBS GLOBAL PROPERTY MLIR_DIALECT_LIBS)
+    get_property(MLIR_CONVERSION_LIBS GLOBAL PROPERTY MLIR_CONVERSION_LIBS)
+    get_property(MLIR_EXTENSION_LIBS GLOBAL PROPERTY MLIR_EXTENSION_LIBS)
+    get_property(MLIR_TRANSLATION_LIBS GLOBAL PROPERTY MLIR_TRANSLATION_LIBS)
+
+    set(MLIR_LIBS
+      MLIRAnalysis
+      MLIRDialect
+      MLIRExecutionEngine
+      MLIRIR
+      MLIRParser
+      MLIRPass
+      MLIRSupport
+      MLIRTransforms
+      MLIRTransformUtils
+
+      ${MLIR_DIALECT_LIBS}
+      ${MLIR_CONVERSION_LIBS}
+      ${MLIR_EXTENSION_LIBS}
+      ${MLIR_TRANSLATION_LIBS}
+    )
+endif()
 
+if (CLANG_LINK_CLANG_DYLIB)
+    set(CLANG_LIBS clang-cpp)
+else()
+    set(CLANG_LIBS
+        clangAST
+        clangASTMatchers
+        clangBasic
+        clangCodeGen
+        clangDriver
+        clangFrontend
+        clangSerialization
+        clangTooling
+    )
+endif()
 #
 # VAST build settings
 #
 
+set(VAST_MAIN_SRC_DIR ${CMAKE_CURRENT_SOURCE_DIR})
+set(VAST_MAIN_INCLUDE_DIR ${VAST_MAIN_SRC_DIR}/include)
+
+set(VAST_LIBRARY_DIR ${CMAKE_BINARY_DIR}/lib)
+set(VAST_TOOLS_DIR ${CMAKE_BINARY_DIR}/tools)
+
+# They are used as destination of target generators.
+set(VAST_RUNTIME_OUTPUT_INTDIR ${CMAKE_CURRENT_BINARY_DIR}/${CMAKE_CFG_INTDIR}/bin)
+set(VAST_LIBRARY_OUTPUT_INTDIR ${CMAKE_CURRENT_BINARY_DIR}/${CMAKE_CFG_INTDIR}/lib${VAST_LIBDIR_SUFFIX})
+if(WIN32 OR CYGWIN)
+  # DLL platform -- put DLLs into bin.
+  set(VAST_SHLIB_OUTPUT_INTDIR ${VAST_RUNTIME_OUTPUT_INTDIR})
+else()
+  set(VAST_SHLIB_OUTPUT_INTDIR ${VAST_LIBRARY_OUTPUT_INTDIR})
+endif()
+
+option(VAST_BUILD_DIALECTS "Build VAST Dialects." ON)
+
+option(VAST_BUILD_CONVERSIONS "Build VAST conversion passes." ON)
+if (NOT VAST_BUILD_DIALECTS AND VAST_BUILD_CONVERSION_PASSES)
+  message(FATAL_ERROR
+    "Conversion passes require dialects to be built."
+    "Set `VAST_BUILD_DIALECTS` option."
+  )
+endif()
+
+option(VAST_BUILD_FRONTEND "Build VAST frontend." ON)
+if (NOT VAST_BUILD_DIALECTS AND VAST_BUILD_FRONTEND)
```

Comment:
`vast-front` also requires conversions/targets (since we need to get llvm somehow).